### DISCUSSION TITLE
What is the fastest/cheapest way to obtain the number of VMs from the inventory?

### DISCUSSION THREAD

**Discussion by atc0005** [vmware/govmomi#3203] [Community Help] [answered]
I'd like to obtain the total count of non-template VMs in the inventory.  In my use case I do not need names or other properties.

In the environment I'm currently testing in there are:

- 529 non-template VMs
- 13 template VMs
- 30 resource pools

I've tested using a finder and using a view. 

Regardless of what settings I try the finder approach is the slowest.

Using a view the fastest approach looks to be specifying just the `name` field, but then I get a total number of VMs which includes templates: 542 

If I retrieve the `config` property also then the time increases substantially, but I can then use the `vm.Config.Template` field to separate the total VMs returned by the retrieve operation into non-template VMs (529) and template VMs (13) to obtain a count of each.

Is there another approach which is faster?

Tips/suggestions are welcome. Thanks in advance!

**Comment by dougm** [comment, marked as answer]
Hey @atc0005 , view is much faster than finder in this case. You can specify the `config.template` property (or `summary.config.template`.  Below is a modify example of [ExampleContainerView_RetrieveWithFilter](https://github.com/vmware/govmomi/blob/c63291aa18123977bb7a143222437d37613c9da6/view/example_test.go#L104)

```go
func ExampleContainerView_vmCount() {
	simulator.Run(func(ctx context.Context, c *vim25.Client) error {
		// quick version of MarkAsTemplate()
		simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine).Config.Template = true

		m := view.NewManager(c)
		kind := []string{"VirtualMachine"}

		v, err := m.CreateContainerView(ctx, c.ServiceContent.RootFolder, kind, true)
		if err != nil {
			log.Fatal(err)
		}

		var vms []types.ObjectContent
		filter := property.Filter{"config.template": true}

		err = v.Retrieve(ctx, kind, filter.Keys(), &vms)
		if err != nil {
			return err
		}

		templates := filter.MatchObjectContent(vms)
		fmt.Printf("%d vms, %d templates\n", len(vms), len(templates))

		return v.Destroy(ctx)
	})
	// Output: 4 vms, 1 templates
}
```

**Reply by atc0005** [reply]
Note for future readers:

As of the v0.35.0 release you will need to swap out `property.Filter` for `property.Match` for the example code to work as intended:

```diff
-filter := property.Filter{"config.template": true}
+filter := property.Match{"config.template": true}
```

See also:

- https://github.com/vmware/govmomi/commit/95aa2571dac212bfb067147a2bcf85b0eeb9afa6

**Comment by atc0005** [comment]
@dougm This is great, thank you!

Thank you also for linking to those examples, I completely overlooked them.

For a managed object like resource pools, would the name property generally be the least expensive or is there a better option (same goal of gathering just the count) ?

**Reply by dougm** [reply]
Yeah `name` or another [ManagedEntity](https://vdc-download.vmware.com/vmwb-repository/dcr-public/d1902b0e-d479-46bf-8ac9-cee0e31e8ec0/07ce8dbd-db48-4261-9b8f-c6d3ad8ba472/vim.ManagedEntity.html) such as `overallStatus`.. would be max 6 chars ([yellow](https://vdc-download.vmware.com/vmwb-repository/dcr-public/d1902b0e-d479-46bf-8ac9-cee0e31e8ec0/07ce8dbd-db48-4261-9b8f-c6d3ad8ba472/vim.ManagedEntity.Status.html)) for each instance.

Maybe we should use that instead of name in the ContainerView.Find method: https://github.com/vmware/govmomi/blob/b5048368213084ecad7bfb014c3ddbafeb600499/view/container_view.go#L116-L117

And using `types.ObjectContent` will use less memory and less reflection magic than the corresponding mo.ResourcePool, etc., types.

On the examples, we should probably make an index somewhere (in the wiki or .md), that'd be easier to find than in the generated godoc.
